File: vec_ops/cpu/vec_axpy.cpp
#include <chrono>
#include <string>
#include <omp.h>
#include <thread>
#include "vec_bench.h"

#define value_t float
#define index_t int

struct vec_axpy : public vec_bench<value_t, index_t>
{

    void benchmark()
    {
        print_bench();

        std::cout << "\nSimulation info: 1d vec axpy\n";

        value_t *x = new value_t[total_size];
        value_t *y = new value_t[total_size];
        value_t a = 1.0;

        #pragma omp parallel for
        for (index_t j = 0; j < total_size; j++)
        {
            x[j] = 1.0;
            y[j] = 0.0;
        }

        loops = 0;
        auto start = std::chrono::high_resolution_clock::now();
        while ((std::chrono::duration_cast<std::chrono::milliseconds>(
                    std::chrono::high_resolution_clock::now() - start)
                    .count()) < 1000.0 * benchtime)
        {
// #pragma omp simd
#pragma omp parallel for
            for (index_t j = 0; j < total_size; j++)
                y[j] += a * x[j];
            loops++;
        }
        auto stop = std::chrono::high_resolution_clock::now();
        duration = 1.0e-3 * std::chrono::duration_cast<std::chrono::milliseconds>(stop - start).count();

        value_t target = (value_t)loops * (value_t)total_size;
        test_result(y, target);
        print_performance();

        delete[] x;
        delete[] y;
    }

    vec_axpy(int narg, char **arg) : vec_bench<value_t, index_t>(narg, arg)
    {
        memory_transfer_per_loop = 3.0 * sizeof(value_t) * double(total_size) /
                                   (1024.0 * 1024.0 * 1024.0);
    }
};

int main(int narg, char **arg)
{
    vec_axpy test(narg, arg);
    test.benchmark();
}

File: vec_ops/jax/vec_axpy.py
import os, sys
import time

sys.path.append(os.path.dirname(os.path.dirname(__file__)))
import vec_bench

import jax
import jax.numpy as jnp

jax.config.update("jax_enable_x64", True)
dtype = jnp.dtype(jnp.float32)
# jax.config.update("jax_platform_name", "cpu")


@jax.jit
def kernel(a, x, y):
    print("Tracing Jax function ...............")
    return a * x + y


class vec_axpy(vec_bench.vec_bench):

    def __init__(self):
        super().__init__()
        self.elem_size = dtype.itemsize
        self.memory_transfer_per_loop = (3.0 * float(self.elem_size) *
                                         float(self.total_size) /
                                         (1024.0 * 1024.0 * 1024.0))

    def benchmark(self):
        self.print_bench()
        print("\nSimulation info: 1d vec axpy\n")

        x = jnp.ones((self.total_size, ), dtype=dtype)
        y = jnp.zeros((self.total_size, ), dtype=dtype)
        self.assert_shape_and_dtype(x)
        self.assert_shape_and_dtype(y)
        a = 1.0

        # warm up
        y = kernel(a, x, y)

        #main loop
        self.loops = 0
        start_time = time.perf_counter()
        while ((time.perf_counter() - start_time) < self.benchtime):
            y = kernel(a, x, y)
            self.loops = self.loops + 1
        y.block_until_ready()
        self.duration = time.perf_counter() - start_time

        target = float(self.loops) * float(self.total_size)
        self.test_result(y, target)
        self.print_performance()


if __name__ == "__main__":
    test = vec_axpy()
    test.benchmark()
